src/cascadia/TerminalControl/UiaTextRange.cpp

```diff
@@ -0,0 +1,194 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "UiaTextRange.hpp"
+#include "TermControlUiaProvider.hpp"
+
+using namespace Microsoft::Terminal;
+
+std::deque<UiaTextRange*> UiaTextRange::GetSelectionRanges(_In_ Microsoft::Console::Render::IRenderData* pData,
+                                                           _In_ IRawElementProviderSimple* pProvider)
+{
+    std::deque<UiaTextRange*> ranges;
+
+    // get the selection rects
+    const auto rectangles = pData->GetSelectionRects();
+
+    // create a range for each row
+    for (const auto& rect : rectangles)
+    {
+        ScreenInfoRow currentRow = rect.Top();
+        Endpoint start = _screenInfoRowToEndpoint(pData, currentRow) + rect.Left();
+        Endpoint end = _screenInfoRowToEndpoint(pData, currentRow) + rect.RightInclusive();
+        UiaTextRange* range = UiaTextRange::Create(pData,
+                                                   pProvider,
+                                                   start,
+                                                   end,
+                                                   false);
+        if (range == nullptr)
+        {
+            // something went wrong, clean up and throw
+            while (!ranges.empty())
+            {
+                UiaTextRangeBase* temp = ranges[0];
+                ranges.pop_front();
+            }
+            throw E_INVALIDARG;

```

Comment:
At least use THROW_HR(E_INVALIDARG) here to let wil do the right thing and log it.